Add new package infrastructure (`epichains` classes and methods)

jamesmbaazam opened the pull request:
This PR addresses several issues:
- Refactored `bpmodels::chain_ll()` to `likelihood()` to close #5 and fix #13. Tests have been delayed to accommodate any changes in name and structure resulting from code review.
- Refactored `bpmodels::chain_sim_susc()` by splitting it into `simulate_tree()` and `simulate_summary()` to close #14. Tests have been delayed to accommodate any changes in name and structure resulting from code review.
- Refactored and renamed `chain_sim_susc()` to `simulate_tree_from_pop()`. This closes #14.  
- Refactored and renamed `chain_ll()` to `likelihood()`. This closes #26.
- Implemented an `epichains` class to partly fix #4. `likelihood()`, which is a refactoring of `chain_ll()` currently does not implement S3 because its use case is currently unknown.  

Essentially, merging this PR will introduce:
- Four simulation functions:
   - `simulate_tree()` - simulates branching processes from a given number of chains.
   - `simulate_tree_from_pop()` - simulates branching processes from a given population size and pre-existing immunity.
   - `simulate_summary()` - simulates a vector of cluster sizes or lengths from branching processes 
   - `likelihood()` - calculates the likelihood/loglihood of observing transmission chains of given sizes or lengths.
- An `epichains` class that stores `data.frame`-like output from `simulate_tree()` and `simulate_tree_from_pop()` with a `chains_tree` attribute and `vector`-like model output from `simulate_summary()` with a `chains_summary` attribute.
- `print()`, `summary()`, and `aggregate()` methods. Note that the `aggregate()` method only works for `epichains` objects with the `chains_tree` attribute.

sbfnk commented on R/simulate.r:
```diff
@@ -55,192 +40,404 @@
+#' simulate_tree_from_pop(pop = 100, offspring_sampler = "nbinom",
+#' mean_offspring = 0.5, disp_offspring = 1.1, serial_sampler = function(x) 3)
+#' @export
+simulate_tree_from_pop <- function(pop,
```
Is the plan to eventially merge this into `simulate_tree` with a `pop` option? I think that might make sense if it doesn't make that function too complex.

sbfnk commented on R/epichains.R:
```diff
@@ -0,0 +1,283 @@
+  if (attributes(x)$chain_type == "chains_tree") {
+    stopifnot(
+      "object does not contain the correct columns" =
+        c("sim_id", "ancestor", "generation", "time") %in%
```
"time" is not strictly necessary - only if used with a serial interval

jamesmbaazam replied:
Fixed in 5d3e108769cc6239b2b28abab7c29f9ae02a7163.

sbfnk commented on R/simulate.r:
```diff
@@ -1,52 +1,37 @@
+#' @param chain_stat_max A cut off for the chain statistic (size/length) being
+#' computed. Results above the specified value, are set to this value.
+#' Defaults to `Inf`.
+#' @param serials_sampler The serial interval generator function; the name of a
```
perhaps rename it something that contains "interval", e.g. `si_sampler` (though see also my other comment on supporting character strings vs. anonymous functions here - I think there might be an argument for just supporting character strings in which case this would be `si_dist`)

sbfnk replied:
I think the result of the discussion (including https://github.com/epiverse-trace/epichains/issues/25#issuecomment-1651715438) is that we want to do this via character strings everywhere instead of functions (though happy to be told otherwise if you read this differently). Perhaps convert my previous comment to an issue for future processing.

sbfnk commented on R/epichains.R:
```diff
@@ -0,0 +1,283 @@
+    )
+
+    # Offer more information to view the full dataset
+    writeLines(sprintf("Use View(<object_name>) to view the full output."))
```
`View` doesn't work well in remote sessions - Perhaps instead have
```suggestion
    writeLines(sprintf("Use as.data.frame(<object_name>) to view the full output."))
```

jamesmbaazam replied:
I've made the change here c0c6197fa0e2e60640140c3380483560a0a4069e.

sbfnk commented on R/epichains.R:
```diff
@@ -0,0 +1,283 @@
+    )
+  }
+
+  structure(out,
```
Does this really need its own S3 class? Do you envision any subsequent operations on this? I think it could make sense to just return a `data.frame`

More generally I also wonder if this function is needed (happy to convinced it is!). It seems the same could be achieved with standard R routines such as e.g.

```r
as.data.frame(x) |>
  dplyr::count(generation)
```

jamesmbaazam replied:
This method is meant to work under the hood of the `plot()` method. The S3 class was to be passed on to `plot()`, but I later removed `plot`, so it makes this look unnecessary. I plan to reimplement the plotting method, after which, this will probably make more sense. The S3 class is used to also store the grouping variable for easy (auto)plotting.

pearsonca commented:
> * this is probably going too far but since you're redisgning already anyway you could consider whether you want to implement this in the line of the more generic modelling functions/packages in R, e.g. have `epichains` (or similar) objects that contain model parameters so one could call `logLik(epichain)` or perhaps even `predict(epichain)` (instead of `simulate`). Again, I'm not at all convinced this is a good idea but I think it might be worth at least thinking about.

Definitely worth thinking about, imo. Gut intuition is that doing some light work would get the basic interface requirements met. Would take a fair bit to get everything properly aligned so that it "just works" with the tools written against the generic interface, but that would be potentially worthwhile.

Re tagging for #25 - happy to contribute code to making the-lambda-licious version actually happen, if that's the direction people want.

sbfnk commented:
> happy to contribute code to making the-lambda-licious version actually happen, if that's the direction people want.

I'm not convinced either way but currently leaning towards using character strings to identify `r`/`d` functions because
1. It means people don't have to provide both sampling and density functions but only one distribution
2. It means we can work with the `truncdist` package (which would otherwise have to be reimplemented - not impossible as it's fairly lightweight - (1) is the stronger argument I think)

Of course the real issue is that the R distribution interface is not particularly clever sensible - we could consider using e.g. the more sensible [distr6](https://alan-turing-institute.github.io/distr6/) which would remove the issue altogether but that would come some overhead.

pearsonca commented:
But itself is a function that takes arguments, yes? So has the same contours as passing any other distribution function + arguments, right?

sbfnk commented:
It's a package that provides distribution-based functions (i.e. `q.../r.../d....`) for truncating arbitrary distributions. Where the arbitrary distributions are specified as character strings and relevant functions pulled from the environment by name.

pearsonca commented:
See update with example; there's no need to reimplement `truncdist` imo. The appropriate interface for interacting with an arbitrary distribution function already handles using something from `truncdist` + whatever-find-the-underlying-function approach.

sbfnk commented:
Ah but the point is that the truncation could come in separately if the user wants [control measures](https://github.com/epiverse-trace/epichains/issues/3#issuecomment-1578443001) or [depletion of susecptibles](https://github.com/epiverse-trace/epichains/blob/82f7d7a7365cc7b2224c5dffcce7414e8c85865d/R/simulate_susceptibles.R#L80). Of course the same could be implemented by the user in just providing a truncated distribution but having these separate would provide convenience arguments.

pearsonca commented:
Hmm - where does `truncdist` do its look up? if it along the normal environment path, couldn't you always just locally (as in, internal to your function) define some generic name for the distribution function, and then hard code that name as the lookup for `truncdist`?

pearsonca commented on R/borel.r:
Convention for documenting the d/r/q/p functions is to use a single rdname.

This is also waaaaaaaaay thin in terms of detail for the distribution, parameters, connection to typical measures (mean, sd, ...), etc.

Probably warrants a devoted issue, rather than tacking on to this already ... dense PR.

pearsonca commented on R/checks.R:
```diff
@@ -0,0 +1,70 @@
+#' @param nchains Number of chains to simulate.
+#'
+#' @keywords internal
+check_nchains_valid <- function(nchains) {
```
warn on non-integer?

jamesmbaazam replied:
This has been fixed here 912ee56c30bc3e531f322c4e10e06b41ebd6eb4c.

pearsonca commented on R/checks.R:
```diff
@@ -0,0 +1,70 @@
+#' @param x An [`epichains`] object
+#'
+#' @keywords internal
+check_chain_tree_attribute <- function(x) {
```
probably better to use the S-whichever architecture for type checking (e.g. `is.epichains` or the like)

pearsonca commented on R/epichains.R:
```diff
@@ -0,0 +1,283 @@
+#' @param x An [`epichains`] object
+#'
+#' @keywords internal
+is_epichains_aggregate_df <- function(x) {
+  if (!inherits(x, "epichains_aggregate_df")) {
+    stop("Object must have class 'epichains_aggregate_df'")
+  }
+}
```
The broader convention for `is_...` functions is to return TRUE / FALSE, not to throw.

Elsewhere, code has `check_...` for the throwing case. Certainly useful to have both, but: generally want to have `check_...` written in terms of `is_` (and per earlier comment, recommend leveraging the type system and using `is.epichains` etc instead), and also exporting `is_...` (definitely something a package consumer might want around).

jamesmbaazam replied:
Thanks, Carl. I've modified the `is_` family to return a logical (`TRUE`), and `check_` to throw custom error messages. For example, here d4c6dc9946247a1b06833d8c2d09a9f0080b8a0d. I will implement `is.epichains` method later.

jamesmbaazam replied:
Thanks, Carl. Good points there. The `is_` functions are now exported and are used to test for various classes and attributes. The `check_` family will throw errors. Examples of changes are in b99f3b06526bd29f0bac56405e179b5733bb1b44, f7c3db5736020549d1c9eeb993774b51ba4d99ca, d4c6dc9946247a1b06833d8c2d09a9f0080b8a0d.

pearsonca commented on R/epichains.R:
```diff
@@ -0,0 +1,283 @@
+#' @return object of class `data.frame`
+#' @author James M. Azam
+#' @export
+head.epichains <- function(x, ...) {
+  utils::head(as.data.frame(x), ...)
+}
+
+#' `tail` method for [`epichains`] class
+#' @param x An [`epichains`] object
+#' @param ... further arguments passed to or from other methods
+#' @importFrom utils tail
+#' @author James M. Azam
+#' @export
+tail.epichains <- function(x, ...) {
+  utils::tail(as.data.frame(x), ...)
+}
```
Is an epichains object also a data.frame object (or could it also be)? If so, I *think* these should be unnecessary?

jamesmbaazam replied:
Yes, it is also a `data.frame`. However, I have been considering implementing a `head` method that only prints from the first known ancestor.

pearsonca commented on R/helpers.R:
```diff
@@ -0,0 +1,87 @@
+#'
+#' @return a function for calculating chain statistics
+#' @keywords internal
+get_chain_statistic_func <- function(chain_statistic) {
+  func <- if (chain_statistic == "size") {
+    rbinom_size
+  } else if (chain_statistic == "length") {
+    rgen_length
+  }
+  return(func)
+}
```
point of taste, but this is just a lookup.

If there needs to be checking of `chain_statistic` makes sense as a function, but its really just a lookup table, this would work just as well:

```
chain_statistic_func <- list(
  size = rbinom_size, length = rgen_length
)
```

jamesmbaazam commented:
Conversation from Slack:

@pearsonca 
hopefully, the rambling commentary on here isn't too scattershot.

I think the upshot from the discussion is that there is likely a way to support the distributions arguments as functions (and probably also backwards-compatible as strings - sniff w/ `is.character()`, if yes, promote to function via `get()`) making implicit explicit: I think those should be functor arguments.

I'm happy to put my keystrokes where my mouth is and help with sorting out how to make that happen

@sbfnk 
Thanks, Carl - not at all too rambling, but I’m still not convinced that (`sampler = rnorm, density = dnorm`)  is preferable to (`dist = "norm"`). I find the second one conceptually more satisfying and potentially (?) less prone to be erroneously used (as we can’t check that sampler and density are consistent with each other).

@pearsonca 
mmmm, seems the (`dist = "mycustom"`) suffers the same consistency problem?

@sbfnk 
yes if using custom then a similar problem exists
though perhaps requiring them to be called r/d with the same name serves as a memo that they have to be the same?

@pearsonca 
agree with preserving the convenience & safety of (`dist = "X"`), but could likely support just (`sampler = rX`) => do the NSE to fish out `dX` in the environment.

probably even (`sampler = pkg::rX`) and fish out from the right environment.


@sbfnk 
Reflecting on this I’m not sure we ever need sampler and density (or, to be precise, log-likelihood as it wouldn’t be e.g. `dnorm` but the log-likelihood of observing chain size x if the offspring distribution is normal) as we use the sampler only to approximate when there is no analytical density. So having these as separate arguments would make that choice explicit which might be a good thing. It would require the user to find out which log-likelihoods are implemented analytically before making the choice but, again, perhaps exposing that is not a bad thing. (edited) 

@pearsonca 
perhaps; I will say that I don't have the overall vocabulary of the package top of mind.


@sbfnk 
It might be good to write down some of the examples / construct a few more with both notations to make sure we’re not missing any obvious design issues


@pearsonca 
but yeah, was about to ask under what circumstances both r/d versions are used at the same time


@sbfnk 
it would require the user to specify the size distribution likelihood instead of the offspring distribution which is possibly less intuitive if more explicit. I’ll have a play with some examples (or you feel free to if you fancy) to share and then we can decide.

> Perhaps; I will say that I don’t have the overall vocabulary of the package top of mind.

I do!

jamesmbaazam commented on .gitignore:
```diff
@@ -31,3 +31,5 @@ rsconnect/
 /Meta/
 /docs/
 .DS_Store
+
+R/test_refactoring.R
```
This has been fixed in cfbf7e5e7015e27018d62bcfeb02796daf32ce70.

jamesmbaazam commented on CITATION.cff:
```diff
@@ -27,23 +27,23 @@ authors:
-url: https://epiverse-trace.github.io/bpmodels/
+preferred-citation:
+  type: manual
+  title: 'epichains: Analysing transmission chain statistics using branching process
+    models'
```
Has been changed to "Simulating and Analysing Transmission Chain Statistics Using Branching Process
    Models" in 8dc51851de927ca9af844f18e6b916becd2c6647.

jamesmbaazam commented on R/epichains.R:
```diff
@@ -0,0 +1,283 @@
+      )
+
+    # print head of the simulation output
+    print(head(x[!is.na(x$ancestor), ]))
```
That's a good point. I actually wonder if it won't be best to sort by `sim_id` then `ancestor`, so that it's easier to see the chains within each simulation. I have implemented that here 6972ed74ef71ea3ac466c714c89525d273ad3d16.

jamesmbaazam commented on R/likelihood_estimation.R:
```diff
@@ -0,0 +1,124 @@
+#' Estimate the (log) likelihood for observed branching processes
+#'
+#' @param chains_observed Vector of sizes/lengths of transmission chains.
+#' @param chain_statistic Statistic given as \code{chains_observed}
```
`chain_statistic` has been changed to `statistic` and the param docs have been inherited from `simulate_tree()`

jamesmbaazam commented on R/likelihood_estimation.R:
```diff
@@ -0,0 +1,124 @@
+#' numbers).
+#' @param nsim_obs Number of simulations if the likelihood is to be
+#' approximated for imperfect observations.
+#' @param log_trans Logical; Should the results be log-transformed? (Defaults
```
This has been fixed.

jamesmbaazam commented on R/likelihood_estimation.R:
```diff
@@ -0,0 +1,124 @@
+#' @param chains_observed Vector of sizes/lengths of transmission chains.
+#' @param chain_statistic Statistic given as \code{chains_observed}
+#' ("size" or "length" of chains).
+#' @param offspring_sampler Offspring distribution: a character string
```
Agreed. I've made the change.

jamesmbaazam commented on R/likelihood_estimation.R:
```diff
@@ -0,0 +1,124 @@
+#' @param log_trans Logical; Should the results be log-transformed? (Defaults
+#' to TRUE).
+#' @param obs_prob Observation probability (assumed constant)
+#' @param chain_stat_max Any chains of this size/length will be
```
This has been addressed.

jamesmbaazam commented on R/likelihood_estimation.R:
```diff
@@ -0,0 +1,124 @@
+#' Estimate the (log) likelihood for observed branching processes
+#'
+#' @param chains_observed Vector of sizes/lengths of transmission chains.
```
> Should this also be able to deal with an epichains object?

That should be doable. I will create an issue for this.

jamesmbaazam commented on R/likelihood_estimation.R:
```diff
@@ -0,0 +1,124 @@
+    size_x <- unlist(sampled_x)
+    if (!is.finite(chain_stat_max)) {
+      chain_stat_max <- max(size_x) + 1
+      }
```
I will fix all indentation issues in a separate PR to avoid inflating this one.

jamesmbaazam commented:
> * this is probably going too far but since you're redisgning already anyway you could consider whether you want to implement this in the line of the more generic modelling functions/packages in R, e.g. have `epichains` (or similar) objects that contain model parameters so one could call `logLik(epichain)` or perhaps even `predict(epichain)` (instead of `simulate`). Again, I'm not at all convinced this is a good idea but I think it might be worth at least thinking about.

I have created an issue for this (#42 ) for a broader discussion.

jamesmbaazam commented:
> * I'm not a fan of the mixture of `vec`/`vect` and think we could do with something more explicit here. Perhaps `_summaries`?

Agreed. I have renamed the function to `simulate_summary()` and changed the attribute it returns for `chain_type` to `chain_summary` here f2cf51dd7c26181b575a127071478a07ff1e0bc7.

> * `estimate_likelihood` doesn't always produce an estimate. If an analytical solution is available then the results is exact. Perhaps this should be just `likelihood`? Would have to rename the file anme too then and distinguish from `likelihoods.R` (which should perhaps `dist_likelihoods.R` or something like that?).

I have made the proposed changes here 3603508c3b0f90298aaedaa698e2bc2404ef4aa3.

sbfnk commented on R/checks.R:
```diff
@@ -0,0 +1,60 @@
+}
+
+
+#' Check if constructed random number generator for offspring exists
```
Isn't this the same as `check_serial_valid` and could it be a single function `check_func_valid` (which we could also use for the `_ll` functions).

jamesmbaazam replied:
This has been converted to an issue here #47.

sbfnk commented on R/epichains.R:
```diff
@@ -0,0 +1,317 @@
+  validate_epichains(x)
+  # Check that the object is of type "chains_tree"
+  if (!is_chains_tree(x)) {
+    stop("object must be an epichains object with 'chains_tree' attribute.")
```
```suggestion
    stop(
      "object must be an epichains object with 'chains_tree' attribute, ",
      "which can be generated using the `simulate_tree` function."
    )
```

sbfnk commented on R/likelihood.R:
```diff
@@ -0,0 +1,113 @@
+  pars <- as.list(unlist(list(...))) ## converts vectors to lists
+
+  ## calculate likelihoods
+  if (exists(ll_func, where = asNamespace("epichains"), mode = "function")) {
```
Should this use a `check_...` function (e.g. the proposed `check_func_valid`)?

jamesmbaazam replied:
This has been converted to an issue here #47.

sbfnk commented on R/likelihood.R:
```diff
@@ -0,0 +1,113 @@
+  ## calculate likelihoods
+  if (exists(ll_func, where = asNamespace("epichains"), mode = "function")) {
+    func <- get(ll_func)
+    likelihoods[calc_sizes] <- do.call(func, c(list(x = calc_sizes), pars))
```
Is the `log` argument ignored in this case?

jamesmbaazam replied:
This has been converted to an issue here #48.

sbfnk commented on R/simulate.r:
```diff
@@ -55,192 +40,404 @@
+#' ## Specifying `serials_sampler` in `simulate_tree()`
+#'
+#' `serials_sampler` must be specified as a named or
 #' [anonymous/inline/unnamed function](https://en.wikipedia.org/wiki/Anonymous_function#R) # nolint
```
As mentioned in the other comment, I think eventually we'll want `serial_dist` or `si_dist` instead of the sampler argument.

sbfnk commented on NEWS.md:
```diff
@@ -1,3 +1,39 @@
+* `simulate_tree()`: simulate transmission trees from a given number of chains.
+* `simulate_tree_from_pop()`: simulate transmission trees from a given number 
+  population size and initial immunity.
+* `simulate_vect()`: simulate a vector of observed transmission chains 
```
will need to be updated

jamesmbaazam replied:
I've fixed that and cleaned up the changelog here 580a98b02552e9c00294d92bcfe6e38f633a9c8b.

sbfnk reviewed (changes requested):
Just a few more questions/comments.

General comments are:
- indentation looks off in many places. Presumably this will be fixed in #40 but it might be good to see if we can re-activate the indentation linter to avoid having to go through separate steps in the future.
- it is difficult to see from this large PR if all functionality from `bpmodels` is correctly transferred - I assume it is but as a general point it might be good to think about adding a few tests that check model validity rather than only functional validity.
- the change here that I'm least sure about is the splitting of `chain_sim`. While cyclomatic complexity and complexity of function arguments have been reduced (which is great) this has come at the expense some code duplication (which is bad). I don't have an immediate satisfactory solution but this would be good to think about further at some point

Once comments have been addressed this can be merged I think - with next steps being fixing the indentation re-instating / adding tests (in separate PRs).

sbfnk commented:
Given how large this PR has grown it might be easier to convert remaining comments to issues and merge as is before addressing them separately.

jamesmbaazam commented on R/checks.R:
```diff
@@ -0,0 +1,60 @@
+#' where \code{\link{rpois}} is the R function to generate Poisson random
+#' numbers).
+#' @keywords internal
+check_offspring_valid <- function(offspring_dist) {
```
This will be fixed here #47.
